pkg/sql/contention/event_store.go

```diff
@@ -348,6 +356,42 @@ func (s *eventStore) resolutionIntervalWithJitter() time.Duration {
 	return jitteredInterval
 }
 
+// contentionKey is used as a key in the eventsAggregated map to group
+// contention events by their characteristics.
+type contentionKey struct {
+	waitingStmtFingerprintID appstatspb.StmtFingerprintID
+	waitingTxnFingerprintID  appstatspb.TransactionFingerprintID
+	blockingTxnFingerprintID appstatspb.TransactionFingerprintID
+	contendedKey             string
+}
+
+func logResolvedEvents(ctx context.Context, events []contentionpb.ExtendedContentionEvent) {
+	eventsAggregated := make(map[contentionKey]time.Duration)
+	for _, event := range events {
+		// Create a struct key instead of a composite string
+		key := contentionKey{
+			waitingStmtFingerprintID: event.WaitingStmtFingerprintID,
+			waitingTxnFingerprintID:  event.WaitingTxnFingerprintID,
+			blockingTxnFingerprintID: event.BlockingTxnFingerprintID,
+			contendedKey:             string(event.BlockingEvent.Key),
+		}
+		// Add the contention time to the existing total for this key combination
+		eventsAggregated[key] += event.BlockingEvent.Duration
+	}
+
+	for key, duration := range eventsAggregated {
+		// Create an AggregatedContentionInfo event for structured logging
+		event := &eventpb.AggregatedContentionInfo{
+			WaitingStmtFingerprintId: redact.SafeString("\\x" + sqlstatsutil.EncodeStmtFingerprintIDToString(key.waitingStmtFingerprintID)),
+			WaitingTxnFingerprintId:  redact.SafeString("\\x" + sqlstatsutil.EncodeTxnFingerprintIDToString(key.waitingTxnFingerprintID)),
+			BlockingTxnFingerprintId: redact.SafeString("\\x" + sqlstatsutil.EncodeTxnFingerprintIDToString(key.blockingTxnFingerprintID)),
+			ContendedKey:             redact.RedactableString(key.contendedKey),
```

Comment:
This is not the way to make a string redactable. You can't just cast a `string` to `RedactableString` because it doesn't necessarily contain redaction markers.

`event.BlockingEvent.Key` is of type `roachpb.Key` which implements the `redact.SafeFormatter` interface. You can use `redact.Sprint(event.BlockingEvent.Key)`. Set the type of `contendedKey` to `RedactableString` and then you just pass it as-is here.

Reply:
ah gotcha, thx for catching that.

Changed to use `redact.Sprint`